### DISCUSSION TITLE
useState is keeping original value and not updating

### DISCUSSION THREAD

**Discussion by kendjr** [vercel/next.js#41623] [Help] [answered]
I am using the Next.js catch all route ('[[...slug]]') to create a page that includes an image based on a parameter in the url. If there are no parameters in the url, the variables are set to a default variables to create the page.
In the [[...slug]].tsx file, I set the two variables to either the parameters in the url, or the default parameters, with an if statement. 

      const router = useRouter();
      let storyType = "";
      let storyBrand = "";
    
      const storyData = router.query.slug;
    
      if (!storyData) {
        storyType = "defaultStoryType";
        storyBrand = "defaultStoryBrand";
      } else {
        storyType = storyData[0];
        storyBrand = storyData[1];
      }

I then send these variables to a component

      return (
        <Fragment>
          <DemoHeader storybrand={storyBrand} storytype={storyType} />
          ....

In the component, I use useState to make sure the image is able to load. If it can't be found, I want to change the source of the image to a default source. The parameters are passed to the component correctly. 

    function DemoImage(props) {
      const { src, alt, width, height } = props;
      console.log("this is the source in demoimage before useState: " + src);
    
      const [imgSrc, setImgSrc] = useState(src);
      console.log("this is imgSrc after useState:     " + imgSrc);
    
      const [imgSrc2, setImgSrc2] = useState("/img/demo/chase/logo.png");
      console.log("this is imgSrc2:    " + imgSrc2);
    
      const handleOnError = () => {
        console.log("could not find imgSrc");
        setImgSrc(
          "https://via.placeholder.com/150x150.png?text=This+is+error+image"
        );
      };
      const handleOnError2 = () => {
        console.log("could not find imgSrc2");
        setImgSrc2("https://via.placeholder.com/150x150.png?text=error+Image2");
      };
    
      return (
        <Fragment>
          <Image
            //{...props}
            width={width}
            height={height}
            src={imgSrc}
            alt={alt}
            onError={handleOnError}
          />
          <Image
            width={width}
            height={height}
            src={imgSrc2}
            alt={alt}
            onError={handleOnError2}
          />
        </Fragment>
      );
    }

The problem is that it goes through the component twice. The first time must be something nextjs does, because it does not even test if the image can be loaded, and variables are not updated with the slug. And the second time through the component it does not reset useState, and therefore appears to use the default values as per the console.log output below. If I reset the useState, it goes through an endless loop.

    FIRST TIME-DOES NOT USE SLUG-IMGSRC SET TO DEFAULT
    This is storyType from slug in slugtsx:  defaultStoryType
    This is the source in demoimage before useState: /img/demo/defaultStoryBrand/logo.png
    This is imgSrc after useState:     /img/demo/defaultStoryBrand/logo.png
    This is imgSrc2:    /img/demo/chase/logo.png
    
    SECOND TIME-HAS SLUG AND CORRECT SRC-DOES NOT SET IMGSRC
    This is storyType from slug in slugtsx:  wsbm
    This is the source in demoimage before useState: /img/demo/chase/logo.png
    This is imgSrc after useState:     /img/demo/defaultStoryBrand/logo.png
    This is imgSrc2:    /img/demo/chase/logo.png
    GET http://localhost:3000/_next/image?url=%2Fimg%2Fdemo%2FdefaultStoryBrand%2Flogo.png&w=640&q=75 400 (Bad Request)	Image (async)	setValueForProperty @ react-dom.development.js?ac89:855	updateDOMProperties @ react-dom.development.js?ac89:9738	updateProperties @ react-dom.development.js?ac89:10143	commitUpdate @ react-dom.development.js?ac89:11047…
    could not find imgSrc 
    This is the source in demoimage before useState: /img/demo/chase/logo.png
    This is imgSrc after useState:     https://via.placeholder.com/150x150.png?text=This+is+error+image

For debugging purposes, I duplicated some of the code using a set variable (imgSrc2) so you can see that it does work correctly if useState is correct.

Can I reset useState without causing an endless loop? Or is there a way to set properly set useState so it uses the updated parameter?

Thanks!

**Comment by gaston-flores** [comment, marked as answer]
You can use `useEffect` to reset your state whenever `src` changes:

```jsx
// in DemoImage
const [imgSrc, setImgSrc] = useState(src);
useEffect(() => {
  setImgSrc(src);
}, [src]);
```

Would this work?

**Reply by kendjr** [reply]
Brilliant! Thanks! 
I don't understand why it "changes", must have something to do with the way it reads the slug, but it must change after useState. 
Either way, this works great.
Thanks!
Ken
